File: java/05-reactor/src/test/java/com/fResult/reactor/ch5_07/SchedulersSubscribeOnTest.java
package com.fResult.reactor.ch5_07;

import static org.junit.jupiter.api.Assertions.assertEquals;

import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.Executors;
import java.util.concurrent.ThreadFactory;
import java.util.concurrent.atomic.AtomicInteger;
import lombok.extern.log4j.Log4j2;
import lombok.val;
import org.junit.jupiter.api.Test;
import reactor.core.publisher.Mono;
import reactor.core.scheduler.Schedulers;
import reactor.test.StepVerifier;

@Log4j2
public class SchedulersSubscribeOnTest {
  @Test
  void subscribeOn() {
    val rsbThreadName = SchedulersSubscribeOnTest.class.getName();
    val map = new ConcurrentHashMap<String, AtomicInteger>();
    val executor = Executors.newFixedThreadPool(5, countingThreadFactory(rsbThreadName, map));
    val scheduler = Schedulers.fromExecutor(executor);
    val integerFlux =
        Mono.just(1)
            .doOnNext(log::info)
            .subscribeOn(scheduler)
            .doFinally(ignored -> map.forEach(this::logKeyValue));

    StepVerifier.create(integerFlux).expectNextCount(1).verifyComplete();
    val atomicInteger = map.getOrDefault(rsbThreadName, new AtomicInteger());
    assertEquals(1, atomicInteger.get());
  }

  private ThreadFactory countingThreadFactory(
      String rsbThreadName, ConcurrentHashMap<String, AtomicInteger> map) {
    return runnable -> {
      Runnable wrapper =
          () -> {
            val key = Thread.currentThread().getName();
            val result = map.computeIfAbsent(key, ignored -> new AtomicInteger());
            result.incrementAndGet();
            runnable.run();
          };
      return new Thread(
          runnable,
          rsbThreadName
              + "-"
              + map.computeIfAbsent(rsbThreadName, k -> new AtomicInteger()).incrementAndGet());
    };
  }

  private void logKeyValue(String key, AtomicInteger value) {
    log.info("Thread: {}, Value: {}", key, value.get());
  }
}

File: kotlin/05-reactor/src/test/kotlin/com/fResult/reactor/ch5_07/SchedulersSubscribeOnTest.kt
package com.fResult.reactor.ch5_07

import org.apache.logging.log4j.LogManager
import reactor.core.publisher.Mono
import reactor.core.scheduler.Schedulers
import reactor.test.StepVerifier
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.Executors
import java.util.concurrent.ThreadFactory
import java.util.concurrent.atomic.AtomicInteger
import kotlin.reflect.jvm.jvmName
import kotlin.test.Test
import kotlin.test.assertEquals

class SchedulersSubscribeOnTest {
  private val log = LogManager.getLogger(SchedulersSubscribeOnTest::class.java)

  @Test
  fun subscribeOn() {
    val rsbThreadName = SchedulersSubscribeOnTest::class.jvmName
    log.info("Test: SchedulersSubscribeOnTest#subscribeOn")
    val map = ConcurrentHashMap<String, AtomicInteger>()
    val executor = Executors.newFixedThreadPool(5, countingThreadFactory(rsbThreadName, map))
    val scheduler = Schedulers.fromExecutor(executor)
    val integer = Mono.just(1).doOnNext(log::info).subscribeOn(scheduler).doFinally { map.forEach(::logKeyValue) }

    StepVerifier.create(integer).expectNextCount(1).verifyComplete()
    val atomicInteger = map.getOrDefault(rsbThreadName, AtomicInteger())
    assertEquals(atomicInteger.get(), 1)
  }

  /**
   * Creates a [ThreadFactory] that produces threads with the given [threadName],
   * and **increments a counter** in the provided [map] every time a thread runs a task.
   *
   * ## Side Effects
   * - Mutates the provided [map] by inserting/updating entries.
   * - Each thread execution increments the associated [AtomicInteger] counter.
   *
   * ## Thread Safety
   * - Assumes [map] is thread-safe (e.g., [ConcurrentHashMap]).
   *
   * @param threadName the name assigned to each created thread
   * @param map a shared, mutable map where thread execution counts will be stored
   * @return a [ThreadFactory] with side effects on [map]
   */
  private fun countingThreadFactory(
    threadName: String,
    map: ConcurrentHashMap<String, AtomicInteger>,
  ): ThreadFactory =
    ThreadFactory { runnable ->
      val wrapper = Runnable {
        val key = Thread.currentThread().name
        val result = map.computeIfAbsent(key) { AtomicInteger() }
        result.incrementAndGet()
        runnable.run()
      }
      Thread(wrapper, threadName)
    }

  fun logKeyValue(key: String, value: AtomicInteger) =
    log.info("Key: [{}], Value: [{}]", key, value)
}
